### PR TITLE
Email addresses in same domain not equal

### LINKED ISSUE #87564: System.Uri considers all email addresses in same domain as equal
### Description

When comparing two instances of `System.Uri` that represent email addresses (`mailto`) scheme, all instances in the same domain are considered equal. The user information in the URI for these addresses seems to be ignored when comparing equality.

### Reproduction Steps

The following failing unit test reproduces the problem:
```c#
    [Fact]
    public void MailtoUriInSameDomainShouldNotBeEqual()
    {
        var uri1 = new Uri( "mailto:first@contoso.com" );
        var uri2 = new Uri( "mailto:second@contoso.com" );
        
        Assert.NotEqual( uri1, uri2 );
    }
```

### Expected behavior

Instances of System.Uri that represent different email addresses should not be equal.

### Actual behavior

Instances of System.Uri that represent email addresses in the same domain are equal, despite differences in the mailbox name.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 7 on macOS Ventura 13.4 and Ubuntu 22.04 jammy (ARM64).

### Other information

In a cursory reading of the `Equals` method code for the type, I did not see any comparison being made to the `UserInfo` property, which is where I believe `System.Uri` holds the mailbox name.

### PR DESCRIPTION
**Pull request by feiyun0112**
Fixes #87564 

The user information in the URI for email addresses in same domain be ignored when comparing equality.

### CODE REVIEW

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 1]
```diff
@@ -1636,6 +1636,11 @@ public override bool Equals([NotNullWhen(true)] object? comparand)
                     return false;
             }
 
+            if (this.Scheme.Equals(UriSchemeMailto))
+            {
+                return this.ToString().Equals(obj.ToString());
```
Comment (gfoidl) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:1641]: Can the string allocations be avoided?

**Comment by danmoseley** [comment]
Is this behavior the same in .NET Framework? 
Would this change break existing code?

**Comment by filipnavara** [comment]
> Is this behavior the same in .NET Framework? Would this change break existing code?

No, it is not. It is a breaking change (in a good direction though). We had a workaround for this for years...

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (feiyun0112) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:1641]: > Can the string allocations be avoided?

how to avoid? please give some advice.

**Review by karelz — APPROVED** [review]

**Comment by karelz** [comment]
Agreed, this is a good technical breaking change. @stephentoub any thoughts here?

**Comment by stephentoub** [comment]
> Agreed, this is a good technical breaking change. @stephentoub any thoughts here?

Without passing judgement on this specific fix, this just seems like a bug, and a bug that's worth fixing.  Like most functional bugs, fixing it could break someone depending on the buggy behavior, but I wouldn't deem this a "breaking change" unless we get a lot of feedback that it's causing problems.

**Comment by danmoseley** [comment]
@marklio  is it possible to say how much usage this api has in nuget?

**Comment by stephentoub** [comment]
> @marklio is it possible to say how much usage this api has in nuget?

It's not just about the API (Equals), which would both almost certainly "a ton" and impossible to measure well because we're talking about an override of object.Equals.  It's more specifically about when the Uri's being compared later are constructed with "mailto" schemas.

**Comment by MihaZupan** [comment]
FWIW, the current behavior [is documented](https://learn.microsoft.com/dotnet/api/system.uri.equals#remarks):

> The [Equals](https://learn.microsoft.com/en-us/dotnet/api/system.uri.equals?view=net-7.0) method compares the two instances without regard to user information ([UserInfo](https://learn.microsoft.com/en-us/dotnet/api/system.uri.userinfo?view=net-7.0)) and fragment ([Fragment](https://learn.microsoft.com/en-us/dotnet/api/system.uri.fragment?view=net-7.0)) parts that they might contain. For example, given the URIs http://www.contoso.com/index.htm#search and http://user:password@www.contoso.com/index.htm, the [Equals](https://learn.microsoft.com/en-us/dotnet/api/system.uri.equals?view=net-7.0) method would return true.

Why this behavior was chosen though, I don't know 🤷‍♂️
While someone out there is likely depending on this being the case for "normal" Uris (e.g. http), I doubt that's the case for `mailto`.

`System.Uri`'s support of the `mailto` scheme is [_limited_](https://github.com/dotnet/runtime/issues/80799#issuecomment-1396083913), such that checking `Equals` on two semantically equivalent mailto addresses may still return false. I doubt we'll ever change that, but this change feels like a good step in the right direction to at least handle the common case better.

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (MihaZupan) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:1641]: This condition is skipping a bunch of edge cases and fast paths.
Can you please move this down to before https://github.com/dotnet/runtime/blob/3fc0f5fd610f73f0397cf40f68bf6e13787a8f2c/src/libraries/System.Private.Uri/src/System/Uri.cs#L1732C9-L1734
and alter which `UriComponents` are selected for comparison.

Something along the lines of
```c#
UriComponents components = UriComponents.HttpRequestUrl;

if (_syntax.InFact(UriSyntaxFlags.MailToLikeUri))
{
    if (!obj._syntax.InFact(UriSyntaxFlags.MailToLikeUri))
        return false;

    components |= UriComponents.UserInfo;
}

string selfUrl = selfInfo.RemoteUrl ??= GetParts(components, UriFormat.SafeUnescaped);
string otherUrl = otherInfo.RemoteUrl ??= obj.GetParts(components, UriFormat.SafeUnescaped);
```

`GetHashCode` should also be updated to include the UserInfo when hashing.

**Comment by marklio** [comment]
Agree with @stephentoub. We can maybe come up with how often Uris are constructed with strings composed with "mailto:" schemes and later compared, but that's not likely to give a picture of what's really going on. Generating "mailto:" Uris is typically done as part of hypertext generation, and not typically used as "data". It seems unlikely that there's alot of code relying on the super odd behavior of "all email addresses from the same domain are equal", because that's not terribly meaningful in a world where gmail and other "large email providers" share the same domain without any other meaningful connection. I'd say we're just as likely to uncover bugs where apps are doing improper equality. That said, I'm happy to try to dig into the data if we think there's value.

**Comment by marklio** [comment]
...although I disagree with not documenting it as a breaking change. :) I think we should.

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 2]
```diff
@@ -1729,9 +1737,19 @@ public override bool Equals([NotNullWhen(true)] object? comparand)
...
+
+            if (_syntax.InFact(UriSyntaxFlags.MailToLikeUri))
+            {
+                if (!obj._syntax.InFact(UriSyntaxFlags.MailToLikeUri))
+                    return false;
+
+                components |= UriComponents.UserInfo;
+            }
+
             // Fragment AND UserInfo are ignored
```
Comment (karelz) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:1750]:
```suggestion
            // Fragment AND UserInfo (for non-mailto URIs) are ignored
```
And move it to the `components` definition above

**Review by karelz — CHANGES_REQUESTED** [review]

**Review by karelz — APPROVED** [review]

**Comment by karelz** [comment]
Thanks @feiyun0112 for your PR!

**Pull request merged**